### PR TITLE
[Y-Cable] Increased unit test coverage of y_cable_helper.py

### PR DESCRIPTION
**Pull request by assrinivasan**
#### Description
Added unit tests for the following methods:

y_cable_wrapper_get_transceiver_info
y_cable_wrapper_get_presence
y_cable_toggle_mux_torA
y_cable_toggle_mux_torB
format_mapping_identifier
update_appdb_port_mux_cable_response_table
update_table_mux_status_for_response_tbl
update_tor_active_side
read_y_cable_and_update_statedb_port_tbl
delete_port_from_y_cable_table
create_tables_and_insert_mux_unknown_entries
check_identifier_presencead_update_mux_table_entry
init_ports_status_for_y_cable
change_ports_status_for_y_cable_change_event
delete_ports_status_for_y_cable
get_firmware_dict
get_muxcable_info
get_muxcable_static_info
post_mux_info_to_db
task_download_firmware_worker

Additionally, 

Added getKeys() functionality to mock swsscommon Table class needed for unit tests
Fixed a minor bug in post_mux_static_info_to_db function

#### Motivation and Context
The motivation for these changes stems from needing to increase the unit test coverage of the y_cable_helper methods to above 50%.

#### How Has This Been Tested?
This was tested by building and running the new unit tests and verifying that they contribute to and increase the code coverage as reflected in the xcvrd_xcvrd_utilities_y_cable_helper_py.html page.

#### Additional Information (Optional)
Moved y_cable_helper test functions to their own file: test_y_cable_helper.py

### CODE REVIEW

File: sonic-xcvrd/tests/test_xcvrd.py [hunk 1]
```diff
@@ -409,7 +455,7 @@ def test_y_cable_toggle_mux_torA_no_port_instance(self):
 
     def test_y_cable_toggle_mux_torA_update_status_exception(self):
 
-        # Maybe I should instantiate a port_instance mock object, give it the toggle_..._tor_a method as an attribute or something?
+        #ASHWIN: TODO: Maybe I should instantiate a port_instance mock object, give it the toggle_..._tor_a method as an attribute or something?
```
Comment (vdahiya12) [inline sonic-xcvrd/tests/test_xcvrd.py:458]: Remove name

File: sonic-xcvrd/tests/test_xcvrd.py [hunk 2]
```diff
@@ -397,6 +420,29 @@ def test_y_cable_wrapper_get_presence_with_platform_chassis(self, mock_chassis):
...
+        mock_object.get_presence = MagicMock(side_effect = NotImplementedError)
+        mock_chassis.get_sfp = MagicMock(return_value = mock_object)
+
+        class mock_sfp_util:
+            def get_presence():
+                return True
+
+        y_cable_platform_sfputil = mock_sfp_util()
+
+        #ASHWIN: TODO: How can I override the local y_cable_platform_sfputil 'NoneType' class object with my mock object containing get_presence method?
```
Comment (vdahiya12) [inline sonic-xcvrd/tests/test_xcvrd.py:437]: remove name, do this throughout the file

**Review by vdahiya12 — CHANGES_REQUESTED** [review]

File: sonic-xcvrd/tests/test_xcvrd.py (hunk 1 above)
Comment (assrinivasan) [reply sonic-xcvrd/tests/test_xcvrd.py:458]: Those items are done, I'll remove all the offending comments in the upcoming commit.

File: sonic-xcvrd/tests/test_xcvrd.py (hunk 2 above)
Comment (assrinivasan) [reply sonic-xcvrd/tests/test_xcvrd.py:437]: Same as above.

File: sonic-xcvrd/tests/test_xcvrd.py [hunk 3]
```diff
@@ -364,6 +657,1448 @@ def test_get_ycable_port_instance_from_logical_port(self):
...
+
+    @patch('xcvrd.xcvrd_utilities.port_mapping.PortMapping.logical_port_name_to_physical_port_list', MagicMock(return_value=[0]))
+    @patch('xcvrd.xcvrd_utilities.y_cable_helper.y_cable_wrapper_get_presence', MagicMock(return_value=False))
+    def test_get_ycable_port_instance_from_logical_port_no_presence(self):
+        with patch('xcvrd.xcvrd_utilities.y_cable_helper.y_cable_port_instances') as patched_util:
+            patched_util.get.return_value = 0
+            port_mapping = PortMapping()
+            instance = get_ycable_port_instance_from_logical_port("Ethernet0", port_mapping)
+
+        assert(instance == PORT_INSTANCE_ERROR )
```
Comment (vdahiya12) [inline sonic-xcvrd/tests/test_xcvrd.py:669]: minor:formatting

File: sonic-xcvrd/tests/test_xcvrd.py [hunk 4]
```diff
@@ -364,6 +657,1448 @@ def test_get_ycable_port_instance_from_logical_port(self):
...
+
+        with patch('xcvrd.xcvrd_utilities.y_cable_helper.y_cable_port_instances') as patched_util:
+            def mock_get():
+                pass
+
+            patched_util.get.return_value = mock_get()
+            port_mapping = PortMapping()
+            instance = get_ycable_port_instance_from_logical_port("E", port_mapping)
+
+        assert(instance == PORT_INSTANCE_ERROR )
```
Comment (vdahiya12) [inline sonic-xcvrd/tests/test_xcvrd.py:684]: minor:formatting

File: sonic-xcvrd/tests/test_xcvrd.py [hunk 5]
```diff
@@ -364,6 +657,1448 @@ def test_get_ycable_port_instance_from_logical_port(self):
...
+                    pass
+
+            patched_util.get.return_value = PortInstanceHelper()
+            port_mapping = PortMapping()
+
+            rc = update_appdb_port_mux_cable_response_table(logical_port_name, port_mapping, asic_index, appl_db, read_side)
+            assert(rc == None)
+
+
+
```
Comment (vdahiya12) [inline sonic-xcvrd/tests/test_xcvrd.py:803]: formatting, you can use install/use autopep8 with autopep8 --maxlinewidth=120 -i to make things easier

File: sonic-xcvrd/tests/test_xcvrd.py (hunk 1 above)
Comment (assrinivasan) [reply sonic-xcvrd/tests/test_xcvrd.py]: Done in latest commit.

File: sonic-xcvrd/tests/test_xcvrd.py (hunk 2 above)
Comment (assrinivasan) [reply sonic-xcvrd/tests/test_xcvrd.py]: Done in latest commit.

File: sonic-xcvrd/tests/test_xcvrd.py (hunk 3 above)
Comment (assrinivasan) [reply sonic-xcvrd/tests/test_xcvrd.py]: Done.

File: sonic-xcvrd/tests/test_xcvrd.py (hunk 4 above)
Comment (assrinivasan) [reply sonic-xcvrd/tests/test_xcvrd.py]: Done.

File: sonic-xcvrd/tests/test_xcvrd.py (hunk 5 above)
Comment (assrinivasan) [reply sonic-xcvrd/tests/test_xcvrd.py]: Ran autopep8 - latest commit reflects that.

**Comment by vdahiya12** [comment]
@assrinivasan can you update the description with name of functions added for coverage

**Comment by assrinivasan** [comment]
> @assrinivasan can you update the description with name of functions added for coverage

Done.

File: sonic-xcvrd/tests/test_xcvrd.py [hunk 6]
```diff
@@ -42,6 +42,20 @@
 global_media_settings = media_settings_with_comma_dict['GLOBAL_MEDIA_SETTINGS'].pop('1-32')
 media_settings_with_comma_dict['GLOBAL_MEDIA_SETTINGS']['1-5,6,7-20,21-32'] = global_media_settings
 
+
```
Comment (vdahiya12) [inline sonic-xcvrd/tests/test_xcvrd.py]: Let's separate out the testcases from xcvrd, you can create a new file test_y_cable.py, move these new tests and other 
ycable tests into that.

File: sonic-xcvrd/tests/test_xcvrd.py [hunk 7]
```diff
@@ -243,77 +257,76 @@ def test_init_port_sfp_status_tbl(self):
...
+                                                                                             'version_nic_active': '1.6MS',
+                                                                                             'version_nic_inactive': '1.7MS',
+                                                                                             'version_nic_next': '1.7MS',
+                                                                                             'version_self_active': '1.6MS',
+                                                                                             'version_self_inactive': '1.7MS',
+                                                                                             'version_self_next': '1.7MS',
+                                                                                             'version_peer_active': '1.6MS',
+                                                                                             'version_peer_inactive': '1.7MS',
+                                                                                             'version_peer_next': '1.7MS'}))
```
Comment (vdahiya12) [inline sonic-xcvrd/tests/test_xcvrd.py:288]: move this to new file

File: sonic-xcvrd/tests/test_xcvrd.py [hunk 8]
```diff
@@ -243,77 +257,76 @@ def test_init_port_sfp_status_tbl(self):
...
+                                                                                                    'tor_peer_lane1_maincursor': '17',
+                                                                                                    'tor_peer_lane1_postcursor1': '7',
+                                                                                                    'tor_peer_lane1_postcursor2': '17',
+                                                                                                    'tor_peer_lane2_precursor1': '7',
+                                                                                                    'tor_peer_lane2_precursor2': '7',
+                                                                                                    'tor_peer_lane2_maincursor': '17',
+                                                                                                    'tor_peer_lane2_postcursor1': '7',
+                                                                                                    'tor_peer_lane2_postcursor2': '17'}))
     def test_post_port_mux_static_info_to_db(self):
```
Comment (vdahiya12) [inline sonic-xcvrd/tests/test_xcvrd.py:330]: move this to new file

File: sonic-xcvrd/tests/test_xcvrd.py [hunk 9]
```diff
@@ -345,41 +410,27 @@ def test_y_cable_wrapper_get_presence(self):
...
-        mux_info_dict['version_peer_inactive'] = '0.7'
-        mux_info_dict['version_peer_next'] = '0.7'
-        mux_info_dict['version_nic_active'] = '0.8'
-        mux_info_dict['version_nic_inactive'] = '0.7'
-        mux_info_dict['version_nic_next'] = '0.7'
-        rc = set_show_firmware_fields("Ethernet0", mux_info_dict, xcvrd_show_fw_res_tbl)
+            rc = y_cable_toggle_mux_torA(1)
 
-        assert(rc == 0)
+        assert(rc == 1)
```
Comment (vdahiya12) [inline sonic-xcvrd/tests/test_xcvrd.py:433]: move all test cases till here to y_cable_helper.py

File: sonic-xcvrd/tests/test_y_cable_helper.py [hunk 10]
```diff
@@ -0,0 +1,1839 @@
+from xcvrd.xcvrd_utilities.port_mapping import *
+from xcvrd.xcvrd_utilities.sfp_status_helper import *
+from xcvrd.xcvrd_utilities.y_cable_helper import *
+from xcvrd.xcvrd import *
```
Comment (vdahiya12) [inline sonic-xcvrd/tests/test_y_cable_helper.py:4]: not required

File: sonic-xcvrd/tests/test_y_cable_helper.py [hunk 11]
```diff
@@ -0,0 +1,1839 @@
...
+import time
+
+if sys.version_info >= (3, 3):
+    from unittest.mock import MagicMock, patch
+else:
+    from mock import MagicMock, patch
+
+from sonic_py_common import daemon_base
+from swsscommon import swsscommon
+from sonic_platform_base.sfp_base import SfpBase
```
Comment (vdahiya12) [inline sonic-xcvrd/tests/test_y_cable_helper.py:17]: not required, please remove

File: sonic-xcvrd/tests/test_y_cable_helper.py [hunk 12]
```diff
@@ -0,0 +1,1839 @@
+from xcvrd.xcvrd_utilities.port_mapping import *
```
Comment (vdahiya12) [inline sonic-xcvrd/tests/test_y_cable_helper.py:1]: check the imports again, and only import what is required

**Review by vdahiya12 — CHANGES_REQUESTED** [review]

File: sonic-xcvrd/tests/test_y_cable_helper.py (hunk 11 above)
Comment (assrinivasan) [reply sonic-xcvrd/tests/test_y_cable_helper.py:17]: Removed in latest commit.

File: sonic-xcvrd/tests/test_y_cable_helper.py (hunk 10 above)
Comment (assrinivasan) [reply sonic-xcvrd/tests/test_y_cable_helper.py:4]: I use functions and variables from all of the imports mentioned here, and removing them broke the unit tests. So I left them alone.

File: sonic-xcvrd/tests/test_y_cable_helper.py (hunk 12 above)
Comment (assrinivasan) [reply sonic-xcvrd/tests/test_y_cable_helper.py:1]: Reason for leaving this in the file is mentioned above.

File: sonic-xcvrd/tests/test_xcvrd.py [hunk 13]
```diff
@@ -42,6 +38,20 @@
 global_media_settings = media_settings_with_comma_dict['GLOBAL_MEDIA_SETTINGS'].pop('1-32')
 media_settings_with_comma_dict['GLOBAL_MEDIA_SETTINGS']['1-5,6,7-20,21-32'] = global_media_settings
 
+
+class helper_logger:
```
Comment (vdahiya12) [inline sonic-xcvrd/tests/test_xcvrd.py:42]: do we really require this, can be removed ?

File: sonic-xcvrd/tests/test_xcvrd.py [hunk 14]
```diff
@@ -42,6 +38,20 @@
...
+
+    def log_warning(self, mock_arg):
+        return True
+
+    def log_debug(self, mock_arg):
+        return True
+
+
 class TestXcvrdScript(object):
     def test_xcvrd_helper_class_run(self):
```
Comment (vdahiya12) [inline sonic-xcvrd/tests/test_xcvrd.py:56]: move this UT also to the other file

**Review by vdahiya12 — CHANGES_REQUESTED** [review]

File: sonic-xcvrd/tests/test_xcvrd.py (hunk 13 above)
Comment (assrinivasan) [reply sonic-xcvrd/tests/test_xcvrd.py:42]: We don't need this in test_xcvrd so I've removed it from there. This provides code coverage for y_cable_helper, so I've left it there.

File: sonic-xcvrd/tests/test_xcvrd.py (hunk 14 above)
Comment (assrinivasan) [reply sonic-xcvrd/tests/test_xcvrd.py:56]: This UT was already duplicated in the other file (Ln 30-40) so we are good there.

File: sonic-xcvrd/tests/test_xcvrd.py (hunk 13 above)
Comment (vdahiya12) [reply sonic-xcvrd/tests/test_xcvrd.py:42]: i still see it in this file.

File: sonic-xcvrd/tests/test_xcvrd.py (hunk 14 above)
Comment (vdahiya12) [reply sonic-xcvrd/tests/test_xcvrd.py:56]: ok, lets remove from here.

**Review by vdahiya12 — CHANGES_REQUESTED** [review]

File: sonic-xcvrd/tests/test_xcvrd.py (hunk 14 above)
Comment (assrinivasan) [reply sonic-xcvrd/tests/test_xcvrd.py]: Done in latest commit.

**Review by vdahiya12 — APPROVED** [review]

**Review by vdahiya12 — APPROVED** [review]

**Pull request merged**